### PR TITLE
Use S_EVENT_KILL and S_EVENT_UNIT_LOST to track unit kills

### PR DESCRIPTION
**Pull request by zhexu14**
This PR addresses Issue 2765 (https://github.com/dcs-liberation/dcs_liberation/issues/2765) and augments the current logic with S_EVENT_UNIT_LOST and S_EVENT_UNIT_KILL events.

### CODE REVIEW

**Comment by DanAlbert** [comment]
Thanks! I'll try to find some time to review in the next few days. (Run the auto formatter though, since the check caught that. We do have pre-commit configured that can run that for you if you set it up)

@Starfire13, you seem to be playing a bunch recently. Any chance you could take this build for a spin and lmk how it plays? Sounds good from the description, but an extra player perspective wouldn't hurt.

**Comment by zhexu14** [comment]
> Thanks! I'll try to find some time to review in the next few days. (Run the auto formatter though, since the check caught that. We do have pre-commit configured that can run that for you if you set it up)
> 
> @Starfire13, you seem to be playing a bunch recently. Any chance you could take this build for a spin and lmk how it plays? Sounds good from the description, but an extra player perspective wouldn't hurt.

Thanks! I've (hopefully) fixed up the formatting now.

**Comment by Starfire13** [comment]
Yeah. I'll take this for a test drive tonight and try it out. 🙂

**Comment by Starfire13** [comment]
Just so I am understanding this right, does this change mean that a plane like this one below will be counted as a kill under aircraft destroyed? I dropped 4 GBU-38s on it. As it is right now (as in the release build), this is not counted as a kill. If this PR fixes that problem, that would be very nice.

Screenshot 2023-04-16 225201

**Comment by Starfire13** [comment]
Ok, this test is being done using the latest dev build, not the PR. Essentially, I want to see what the current build is doing wrong, before I test this PR. I still don't quite get what the change does.

So I bombed a couple of parked L-39s with GBU-38s.
Screenshot 2023-04-16 234812

DCS tells me I got three kills.
Screenshot 2023-04-16 235225

Liberation similarly tells me I got three kills.
Screenshot 2023-04-16 235239

Is this an issue with only certain types of munitions? I can't seem to replicate the "killed but aircraft is still visible" thing.

**Comment by zhexu14** [comment]
From what I can tell, there's two issues at play here:
1) In DCS, a unit set to "Uncontrolled" requires alot of firepower to be thrown at it for a kill event to be triggered. Mavericks/GAU-8 don't seem to be enough. A GBU-10 is sufficient. It seems to be very target/weapon specific. I'm keen to work on this issue, and have an idea on how it could be fixed, but that's probably a different PR. 
2) It appears that from https://github.com/dcs-liberation/dcs_liberation/discussions/2764, DCS might not create a Crash event in all circumstances, which meant that Liberation won't register a kill. TBH I've been trying to reproduce this as well without much success (I'm not the person who originally reported the issue). However, I'll keep trying and will let you know -- one thing I'll try is using a BK90, which is what the original reporter was using, but will have to learn the Viggen first :)

**Comment by Starfire13** [comment]
Yes. Parked aircraft are a pain to kill with anything apart from a big bomb. Mavericks, rockets, guns, and cluster munition all tend not to get kills. The planes just burn for a bit in pieces and eventually despawn rather than generate a kill event.

**Comment by Starfire13** [comment]
Ok, I did a test with JSOW-As to see if this is a problem that occurs with cluster munitions. Once again, I am not getting any kills that DCS registers but Liberation does not. That isn't to say this PR doesn't fix anything (it makes sense to me to use lost rather than crash, theoretically). I just can't replicate a scenario (i.e. parked plane killed in DCS but not counted in Liberation) where the release build would need this fix.

JSOW-A hits a bunch of planes. Gets 1 kill (seen on BDA)
Screenshot 2023-04-17 103554

This is reflected in end of mission screen in DCS
Screenshot 2023-04-17 104251

And also in Liberation
Screenshot 2023-04-17 104313

**Comment by zhexu14** [comment]
@Starfire13  I think I've come up with two scenarios which will reproduce the error. 

So far I've tested these two scenarios in toy missions in DCS and confirmed that DCS will generate an S_EVENT_KILL event but *not* an S_EVENT_CRASH event that the develop branch of Liberation relies on to track aircraft kills. 

Scenario 1: hit an aircraft that is taxiing enough that the pilot ejects, but not enough that destroys the aircraft completely. A short burst of a GAU-8 seems to do the trick. DCS will show the "xxx destroyed" message but the aircraft never crashes. I've attached a sample track *scenario1a.trk* and DCS debrief *scenario1a.log* as examples.

However, in the same scenario, if the aircraft is hit with something bigger and destroyed completely (I tried a 2000lb bomb), a S_EVENT_CRASH event is generated and Liberation will track the kill correctly. I've attached a sample track / debrief log as *scenario1b.trk* and  *scenario1b.log*

Also, the aircraft has to be controlled i.e. performing a mission, but still on the ground. In my previous tests, all the aircraft were set up in the mission editor as uncontrolled.

Scenario 2: an aircraft is hit at high altitude. The S_EVENT_KILL is generated immediately, but the S_EVENT_CRASH is generated only when the aircraft hits the ground. If the player exits the mission after the aircraft is hit but before it crashes to the ground, Liberation will not catch the kill. I've attached an example as  *scenario2.trk* and  *scenario2.log*, showing that roughly a minute elapses between when the IL78 generates the  S_EVENT_KILL and the  S_EVENT_CRASH events.

[scenarios.zip](https://github.com/dcs-liberation/dcs_liberation/files/11250969/scenarios.zip)

**Comment by DanAlbert** [comment]
Thanks for the research 👍 

No one has found any cases where crash is emitted but kill isn't, right? If not, I'm convinced by the premise (I'll have time to review tonight unless work goes weird).

File: game/debriefing.py [hunk 1]
```diff
@@ -109,16 +109,43 @@ class StateData:
...
+
+        # process killed units from S_EVENT_UNIT_LOST & S_EVENT_KILL
+        killed_units = clean_unit_list(data["unit_lost_events"] + data["kill_events"])
+        for unit in killed_units:  # organize killed units into aircraft vs ground
+            if unit_map.flight(unit) is not None:
+                killed_aircraft.append(unit)
+            else:
+                killed_ground_units.append(unit)
+
+        # process killed units from S_EVENT_CRASH
```
Comment (DanAlbert) [inline game/debriefing.py]: Is there any reason to continue listening for this?

File: game/debriefing.py [hunk 2]
```diff
@@ -109,16 +109,43 @@ class StateData:
...
+        for unit in data["crash_events"]:  # no need to clean crash events
+            if unit not in killed_aircraft:
+                killed_aircraft.append(unit)
+
+        # process killed units from S_EVENT_DEAD
```
Comment (DanAlbert) [inline game/debriefing.py]: Or this?

If there's a reason to listen to all four of `S_EVENT_UNIT_LOST`, `S_EVENT_KILL`, `S_EVENT_DEAD` and `S_EVENT_CRASH`, could you explain those in a comment here?

File: game/debriefing.py [hunk 3]
```diff
@@ -109,16 +109,43 @@ class StateData:
...
+        # process killed units from S_EVENT_UNIT_LOST & S_EVENT_KILL
+        killed_units = clean_unit_list(data["unit_lost_events"] + data["kill_events"])
+        for unit in killed_units:  # organize killed units into aircraft vs ground
+            if unit_map.flight(unit) is not None:
+                killed_aircraft.append(unit)
+            else:
+                killed_ground_units.append(unit)
+
+        # process killed units from S_EVENT_CRASH
+        for unit in data["crash_events"]:  # no need to clean crash events
```
Comment (DanAlbert) [inline game/debriefing.py]: > no need to clean crash events

Why not? DCS presumably won't emit a duplicate crash event, but don't you need to dedup everything on L147-148? DCS might not emit a duplicate crash, but might it emit a kill, crash, dead, and unit lost all for one unit?

If I'm right about that, I think what you're better off making `killed_aircraft` and `killed_ground_units` into `set[str]` and just doing the list conversion at the end?

If not, could you expand the comment explaining why that doesn't matter?

**Review by DanAlbert — CHANGES_REQUESTED** [review]
LGTM in general, but the comments need expanding (and depending on the answer to one of those, I may have spotted a bug)

File: game/debriefing.py (hunk 1 above)
Comment (zhexu14) [reply game/debriefing.py]: Yes, we definitely need to continue listening to these events to handle scenarios where either the player or AI flies into the ground without being killed e.g. CFIT.

File: game/debriefing.py (hunk 3 above)
Comment (zhexu14) [reply game/debriefing.py]: Good point, we do need to deduplicate all event types. Have updated the code to do this.

File: game/debriefing.py (hunk 2 above)
Comment (zhexu14) [reply game/debriefing.py]: The idea behind listening to all events is that it's not entirely clear what logic DCS uses to generate these events, and there's no guarantee that this logic won't change over time. To have a robust kill detection mechanism, the approach was to log all relevant events even though it results in duplicates. I've added a comment to this effect.

If we want a minimal set of events, then based on my testing so far, we should capture at least:
- S_EVENT_KILL
-  S_EVENT_DEAD (but only for recording the location of destroyed ground objects)
-  S_EVENT_CRASH

**Review by DanAlbert — APPROVED** [review]
LGTM now, thanks!

**Pull request merged**